tox.ini

```diff
@@ -36,15 +26,35 @@ changedir = .tmp/{envname}
 #
 #     tox -l -v
 #
+description =
+    run tests
+    devdeps: with the latest developer version of key dependencies
+    oldestdeps: with the oldest supported version of key dependencies
+    online: that require remote data
+    hypothesis: using hypothesis (both offline and online)
+    figure: runs the figure test suite.
+setenv =
+    MPLBACKEND = agg
+    COLUMNS = 180
+    PYTEST_COMMAND = pytest --pyargs ndcube {toxinidir}/docs --cov=ndcube --cov-config={toxinidir}/setup.cfg
+# Pass through the following environemnt variables which may be needed for the CI
+passenv = HOME WINDIR LC_ALL LC_CTYPE CC CI TRAVIS
 extras =
     test
 deps =
     pytest-xdist
-    devdeps: git+https://github.com/sunpy/sunpy
     devdeps: git+https://github.com/astropy/astropy
+    devdeps: git+https://github.com/sunpy/sunpy
+
+    # Figure tests need a tightly controlled environment
+    figure: matplotlib==3.3.2
+    figure-!devdeps: astropy==4.1.0
```

Comment:
What does ```-!``` mean?

Reply:
Not that factor so `-figure` would match but `-figure-devdeps` would not.